index.js

```diff
@@ -128,46 +153,65 @@ class S3Adapter {
       const serializedTags = serialize(options.tags);
       params.Tagging = serializedTags;
     }
-    return this.createBucket().then(() => new Promise((resolve, reject) => {
-      this._s3Client.upload(params, (err, response) => {
-        if (err !== null) {
-          return reject(err);
-        }
-        return resolve(response);
-      });
-    }));
+    await this.createBucket();
+    const command = new PutObjectCommand(params);
+    const response = await this._s3Client.send(command);
+    return response;
   }
 
-  deleteFile(filename) {
-    return this.createBucket().then(() => new Promise((resolve, reject) => {
-      const params = {
-        Key: this._bucketPrefix + filename,
-      };
-      this._s3Client.deleteObject(params, (err, data) => {
-        if (err !== null) {
-          return reject(err);
-        }
-        return resolve(data);
-      });
-    }));
+  async deleteFile(filename) {
+    const params = {
+      Bucket: this._bucket,
+      Key: this._bucketPrefix + filename,
+    };
+    await this.createBucket()
+    const command = new DeleteObjectCommand(params);
+    const response = await this._s3Client.send(command)
+    return response;
   }
 
   // Search for and return a file if found by filename
   // Returns a promise that succeeds with the buffer result from S3
-  getFileData(filename) {
-    const params = { Key: this._bucketPrefix + filename };
-    return this.createBucket().then(() => new Promise((resolve, reject) => {
-      this._s3Client.getObject(params, (err, data) => {
-        if (err !== null) {
-          return reject(err);
-        }
-        // Something happened here...
-        if (data && !data.Body) {
-          return reject(data);
-        }
-        return resolve(data.Body);
+  async getFileData(filename) {
+    const params = {
+      Bucket: this._bucket,
+      Key: this._bucketPrefix + filename,
+    };
+    await this.createBucket()
+    const command = new GetObjectCommand(params);
+    const response = await this._s3Client.send(command);
+    if (response && !response.Body) throw new Error(response);
+
+    const buffer = await responseToBuffer(response);
+    return buffer;
+  }
+
+  // Exposed only for testing purposes
+  getSignedUrlSync(client, command, options) {
+    let isDone = false;
+    let signedUrl = '';
+    let error = null;
+
+    getSignedUrl(client, command, options)
+      .then((url) => {
+        signedUrl = url;
+        isDone = true;
+      })
+      .catch((err) => {
+        error = err;
+        isDone = true;
       });
-    }));
+
+    // Block the event loop until the promise resolves
+    while (!isDone) {
+      deasync.sleep(100); // Sleep for 100 milliseconds
```

Comment:
100ms seems quite long, what is this value based on?

Reply:
The 100ms value was an estimate to prevent CPU overload from a tight loop, as there's no specific AWS documentation on the exact timing needed for this function. Reducing it to 10ms might be possible. Additionally, I suggest opening an issue on parse-server to make the getFileLocation function async, eliminating the need for the deasync library altogether.

Reply:
I reduced it to 10ms in the new commit

Reply:
If we estimate the response time for an intra-region call from EC2 to S3 (i.e. Parse Server also running on AWS) to be ~10-100ms, then a 100ms sleep would on avg. double the response time. Maybe setting to 50ms could be a good compromise.

Reply:
Thank you for the feedback! I don't currently have a Parse Server test setup, but I can certainly set one up. It might take some time, but I'm definitely up for the task. I'm happy to work on this without needing the bounty.

Could you please create an issue on the Parse Server repository for this?

Reply:
Done, see https://github.com/parse-community/parse-server/issues/9268.

Reply:
Thank you! 🙏 I'll get started on it. 🚀